chore: Simplify Makefile Architecture (#832)
kyma-project/telemetry-manager: docs/contributor/arch/009-makefile-architecture.md

# 9. Makefile: Simplified Architecture

Date: 2024-28-02

## Status

Proposed

## Context: Cluttered Makefile Targets

> Makefile targets are currently cluttered, unorganized, and redundant (see the dependency graph bellow). Targets are not parametrized, thus resulting in multiple make targets serving the same scope (e.g. running e2e tests). Targets are also used simultaneously by developers/contributors, as well as within GitHub actions and shell scripts, which makes dependencies even more intertwined. These factors result in a complex environment that is becoming hard to manage by the developers.

Reviewer: ```suggestion
Makefile targets are currently cluttered, unorganized, and redundant (see the dependency graph below). Targets are not parametrized, thus multiple make targets serve the same scope (for example, running e2e tests). Targets are also used simultaneously by developers and contributors, as well as within GitHub actions and shell scripts, which makes dependencies even more intertwined. These factors result in a complex environment that is becoming hard to manage by the developers.
```

![Makefile Dependency Graph (After)](../assets/makefile-graph-before.png)

## Decision

> After analyzing repositories that have similar configurations as telemetry-manager (e.g. Kubebuilder default configuration, Kyma modules repositories, other Go repositories), the following decisions have been made (see image bellow):

Reviewer: ```suggestion
After analyzing repositories that have similar configurations as Telemetry Manager (such as Kubebuilder default configuration, Kyma modules repositories, and other Go repositories), the following decisions have been made:
```

![Makefile Archtiecture](../assets/makefile-architecture.jpg)

> 1. Split general Makefile into sub-makefiles:

Reviewer: excellent use of ordered list (subsequent steps) and unordered list items, even with correct indentation!

   - `Makefile`: root makefile, containing targets related to the development, building, deployment, and testing flow of the codebase. Includes `provision.mk` and `dependencies.mk`
   - `provision.mk`: contains all targets that (de-)provision a certain environment (e.g. k3d or gardener).
   - `dependencies.mk`: contains all targets that install build dependencies (e.g. kyma, ginkgo, etc.).

> 2. Implement .env file, that stores environment variables accessible from both makefiles and shell scripts:

Reviewer: ```suggestion
2. Implement .env file that stores environment variables accessible from both makefiles and shell scripts:
```
